embedded-uart-common/sample-implementations/arduino-mkrzero-multiple-uart/sensirion_uart_implementation.cpp

```diff
@@ -0,0 +1,156 @@
+ *
+ * * Neither the name of Sensirion AG nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+// needed for delay() routine
+#include "wiring_private.h"  // pinPeripheral() function
+#include <Arduino.h>
+
+#include "sensirion_uart.h"
+
+#define BAUDRATE 115200  // baud rate of SPS30
+#define PIN_UART_2_RX 7
+#define PIN_UART_2_TX 6
+#define PIN_UART_3_RX 29
+#define PIN_UART_3_TX 28
+
+Uart Serial2(&sercom3, 7, 6, PAD_SERIAL1_RX, PAD_SERIAL1_TX);
+Uart Serial3(&sercom4, 29, 28, PAD_SERIAL1_RX, PAD_SERIAL1_TX);
+
+void SERCOM3_Handler() {
+    Serial2.IrqHandler();
+}
+
+void SERCOM4_Handler() {
+    Serial3.IrqHandler();
+}
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+static Uart *const ports[] = {&Serial2, &Serial3};
+static uint8_t cur_port = 0;
+
+/**
+ * sensirion_uart_select_port() - select the UART port index to use
+ *                                THE IMPLEMENTATION IS OPTIONAL ON SINGLE-PORT
+ *                                SETUPS (only one SPS30)
+ *
+ * Return:      0 on success, -1 if the selected port is invalid
+ */
+int16_t sensirion_uart_select_port(uint8_t port) {
+    if (port >= (sizeof(ports) / sizeof(ports[0])))
+        return -1;
+    cur_port = port;
+    return 0;
+}
+
+/**
+ * sensirion_uart_open() - initialize UART
+ *
+ * Return:      0 on success, -1 if the selected port is invalid
+ */
+int16_t sensirion_uart_open() {
+    switch (cur_port) {
+        case 0:
+            Serial2.begin(BAUDRATE);
+            pinPeripheral(PIN_UART_2_RX, PIO_SERCOM_ALT);
+            pinPeripheral(PIN_UART_2_TX, PIO_SERCOM_ALT);
+            return 0;
+
+        case 1:
+            Serial3.begin(BAUDRATE);
+            pinPeripheral(PIN_UART_3_RX, PIO_SERCOM_ALT);
+            pinPeripheral(PIN_UART_3_TX, PIO_SERCOM_ALT);
+            return 0;
+
+        default:
+            return -1;
+    }
+}
+
+/**
+ * sensirion_uart_close() - release UART resources
+ *
+ * Return:      0 on success, -1 if the selected port is invalid
+ */
+int16_t sensirion_uart_close() {
+    switch (cur_port) {
```

Comment:
to be consistent with i2c we should also use
```cpp
ports[cur_port]->end()
```
instead of switch

Reply:
Didn't think about i2c, but in this case I'm more worried about remaining consistent with `sensirion_uart_open` where I use `switch` and hardcode the serial interface. That was the reason to do it like this.